src/main/java/xbot/common/subsystems/vision/AprilTagVisionSubsystemCamera.java

```diff
@@ -0,0 +1,162 @@
+package xbot.common.subsystems.vision;
+
+import edu.wpi.first.apriltag.AprilTagFieldLayout;
+import edu.wpi.first.math.VecBuilder;
+import edu.wpi.first.math.geometry.Pose3d;
+import edu.wpi.first.wpilibj.Alert;
+import org.littletonrobotics.junction.Logger;
+import xbot.common.advantage.DataFrameRefreshable;
+import xbot.common.properties.DoubleProperty;
+import xbot.common.properties.PropertyFactory;
+
+import java.util.LinkedList;
+import java.util.List;
+
+public class AprilTagVisionSubsystemCamera implements DataFrameRefreshable {
+    private final AprilTagVisionIO io;
+    final VisionIOInputsAutoLogged inputs;
+    private final String logPath;
+    private final Alert disconnectedAlert;
+    private final AprilTagFieldLayout aprilTagFieldLayout;
+
+    // Basic filtering thresholds
+    private final DoubleProperty maxAmbiguity;
+    private final DoubleProperty maxZError;
+
+    // Standard deviation baselines, for 1 meter distance and 1 tag
+    // (Adjusted automatically based on distance and # of tags)
+    private final DoubleProperty linearStdDevBaseline;
+    private final DoubleProperty angularStdDevBaseline;
+
+    // Multipliers to apply for MegaTag 2 observations
+    private final DoubleProperty linearStdDevMegatag2Factor;
+    private final DoubleProperty angularStdDevMegatag2Factor;
+
+    // Standard deviation multipliers for each camera
+    // (Adjust to trust some cameras more than others)
+    private final DoubleProperty cameraStdDevFactor;
+
+    private final List<Pose3d> tagPoses = new LinkedList<>();
```

Comment:
⭐ something I've been wondering about, this list doesn't tell us the tag id of each tag, should we keep a hashmap maybe or some other structure so we know what tag id it was? Eg we're looking for a particular tag.

Reply:
You *could* just compare with the AprilTagFieldLayout. I've added some helpers to at least indicate if a tag is visible land which camera can see it.